### PR TITLE
DateRange component - adds clear button and improves a11y

### PR DESCRIPTION
**Pull request by getdave**
Previously there was no easy way to completely clear the currently selected dates and have this persisted. Also there were a few minor a11y concerns to do with decorative icons and missing text within UI buttons.

#### Changes proposed in this Pull Request

* Adds a "Clear" button to the `DateRangeTrigger` which _completely_ deselects _all_ Dates when clicked. This allows user to remove a selected range of dates which is useful in 99% of scenarios (think filtering a set of photos by a date range - you might want to just say "clear all the Date filters")
* Renames info area terminology within the `Popover` from “clear” to “reset” to better convey the intent that it merge resets the dates to the previously selected dates
* Improves a11y by utilising `ScreenReaderText` component and add appropriate `aria-` attrs

These changes are required (and were discovered) as a result of ongoing work in:

* https://github.com/Automattic/wp-calypso/pull/30323/
* https://github.com/Automattic/wp-calypso/pull/25762

#### Testing instructions

* Goto http://calypso.localhost:3000/devdocs/design/date-range - does it behave as you'd expect by default? You should be able to completely clear any selected date ranges returning the component to a completely clean state without any dates.
* Does adding `showClearBtn: false` as a prop to the parent component hide the `clear` button?
* Do the tests pass? `npm run test-client:watch client/components/date-range/test`

#### Questions

* Should the act of clearing the dates trigger the callback props `onDateSelect` and `onDateCommit`? If not then you need to remember to manually wire up a handler for these in the component that is consuming the `DateRange` component.

#### Screencaptures

screen capture on 2019-01-24 at 12-36-00

### CODE REVIEW

File: client/components/date-range/index.js [hunk 1]
```diff
@@ -638,12 +658,16 @@ export class DateRange extends Component {
 		} );
 
 		const triggerProps = {
+			startDate: this.state.startDate,
+			endDate: this.state.endDate,
 			startDateText: this.toDateString( this.state.startDate ),
 			endDateText: this.toDateString( this.state.endDate ),
 			buttonRef: this.triggerButtonRef,
 			onTriggerClick: this.togglePopover,
+			onClearClick: this.clearDates,
```
Comment (spen) [inline client/components/date-range/index.js]: I think we may need an `onClear` prop or similar for controlled usages. If we were to connect this to redux I don't think we'd have a clean way of updating a state slice via a connected action.
I may be wrong, but I think we'd lose the ability to do what https://github.com/Automattic/wp-calypso/pull/30323/files#diff-f30509e27f891d5bbbf525f40a3e6f67R272 does here without the additional prop

**Comment by getdave** [comment]
@keoshi Thanks for feedback. Very much appreciated.

> One thing I noticed is that the clear button doesn't have the same styling on :focus as the date range 

Agreed. Needs to be addressed. I think we should raise an `Issue` for this no?

> Last thing I noticed while testing this is that once a range is selected, I'd expect a click outside of the popover to dismiss it but keep the selected range.

This is intentional behaviour. The guiding principle of the picker was only "commit" the dates when the user clicks "Apply". Anything else should be assumed that they didn't want to commit but accidentally clicked away. If as a group we all feel this is misleading then I'm more than happy to be lead by the majority here. Feels like something that needs user testing. Not sure if/how we facilitate such things.

> we should merge it and then tackle the other issues in separate PRs

Thanks I agree.

File: client/components/date-range/index.js [hunk 2]
```diff
@@ -396,6 +406,21 @@ export class DateRange extends Component {
 		} );
 	};
 
+	/**
+	 * Fully clears all dates to empty values
+	 * Affecting saying "get rid of all dates"
```
[the remark this answers is not in the document]
Comment (getdave) [reply client/components/date-range/index.js]: You are right! I have a bad habit of doing things like this...

File: client/components/date-range/test/index.js [hunk 3]
```diff
@@ -202,6 +202,38 @@ describe( 'DateRange', () => {
 
 			expect( popover.props().isVisible ).toBe( false );
 		} );
+
+		test( 'should reset Dates on trigger clean btn click', () => {
```
[the remark this answers is not in the document]
Comment (getdave) [reply client/components/date-range/test/index.js]: Yep. Typo here.

File: client/components/date-range/index.js (hunk 1 above)
Comment (getdave) [reply client/components/date-range/index.js]: This appears to relate to the question I left in the PR description:

> Should the act of clearing the dates trigger the callback props onDateSelect and onDateCommit? If not then you need to remember to manually wire up a handler for these in the component that is consuming the DateRange component.

Is this the same issue? Ie: we need a way to propagate the clearing of state outwards?

**Comment by keoshi** [comment]
> I think we should raise an `Issue` for this no?

Yes, please! Do you want me to handle that?

> This is intentional behaviour. (...) Feels like something that needs user testing. 

True, and I agree that this should be tested with users. My point of view was tied to the fact that we're showing “reset **selected** dates” (emphasis mine) in the dialog once the end date is clicked. So it seems that they become finalized at that point, which is not true unless you click the apply button to confirm it.

File: client/components/date-range/index.js (hunk 1 above)
Comment (spen) [reply client/components/date-range/index.js]: > Is this the same issue?

Ahh yes, I got a bit confused there but that does look to be the same issue, just in a slightly different manner. With an `onClear` prop there's a particular and explicit action taking place. By calling `onDateCommit` with `null` values (effectively the same) we're communicating up in an implicit way.
With this approach we, on the one hand, are avoiding adding another prop, but on the other hand are being less explicit - the person using this component would have to add extra checking logic in to their handler, something like so:

```js
onDateCommit = ( startDate, endDate ) => {

  if ( ! startDate && ! endDate ) {
    // infer a sort of 'clearDates' event and run relevant logic
    return;
  }
  // continue with regular onCommitDates logic
}
```

That said we might end up getting in to issues by running `onDateCommit` and `onClear` together

I'm not sure which way is best to be honest. My initial thought was the explicit `onClear` but now I'm leaning the way of just calling `onDateCommit`.

> We need a way to propagate the clearing of state outwards?

For sure, one way or another

**Comment by getdave** [comment]
> Yes, please! Do you want me to handle that?

@keoshi would be hugely appreciated!

> showing “reset selected dates” (emphasis mine) in the dialog once the end date is clicke

Would just removing the word `selected` be of a little help here. I know it's not a total solution but still...

File: client/components/date-range/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (getdave) [reply client/components/date-range/index.js]: Fixed

File: client/components/date-range/test/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (getdave) [reply client/components/date-range/test/index.js]: Fixed

**Comment by keoshi** [comment]
Still, on this point:

> The guiding principle of the picker was only "commit" the dates when the user clicks "Apply". Anything else should be assumed that they didn't want to commit but accidentally clicked away.

I pinged my team for thoughts on the “click outside the popover” question and those who weighed in agreed that if both:

- a start date is selected;
- and end date is also selected;

then we should confirm the selection when dismissing the popover.

@rickybanister had a particularly good point:

> yes, even just to defensively help out confused people

Which I think makes a lot of sense if you consider someone that just spent 10~20 seconds making a careful selection and then loses said selection when clicking outside. It’s easy to clear it out (especially with the button @getdave added on this PR), but _not super easy to go back and redo the selection_.

I’d rather have a more conservative approach and confirm their selection in that case.

**Comment by getdave** [comment]
> I’d rather have a more conservative approach and confirm their selection in that case.

Really good feedback here. Great to have some more input.

I've adjusted the Component so that it will _only_ revert the dates if you click `Cancel`. All other scenarios result in the date being committed and persisted.

I've left the `Apply` button in place both for UX and a11y reasons as I believe people prefer to see a clear button that applies a selection (a la normal forms).

@keoshi Does this resolve the issue (see below):

screen capture on 2019-01-25 at 09-40-50

**Comment by keoshi** [comment]
@getdave Yes, that's brilliant! Thank you for making that change! 🙌

File: client/components/date-range/index.js (hunk 1 above)
Comment (getdave) [reply client/components/date-range/index.js]: I've gone with triggering `onDateCommit` when the dates are cleared. Take your point about consuming requiring additional logic but I think they will just need to handle that.

File: client/components/date-range/index.js [hunk 4]
```diff
@@ -411,14 +417,22 @@ export class DateRange extends Component {
...
 	clearDates = () => {
-		this.setState( {
-			startDate: null,
-			endDate: null,
-			staleStartDate: null,
-			staleEndDate: null,
-			textInputStartDate: '',
-			textInputEndDate: '',
-		} );
+		this.setState(
```
[the remark this answers is not in the document]
Comment (getdave) [reply client/components/date-range/index.js]: No. Resolved.

File: client/components/date-range/index.js [hunk 5]
```diff
@@ -565,7 +579,7 @@ export class DateRange extends Component {
 	renderPopover() {
 		const headerProps = {
 			onApplyClick: this.commitDates,
-			onCancelClick: this.closePopover,
+			onCancelClick: this.closePopoverandRevert,
```
[the remark this answers is not in the document]
Comment (getdave) [reply client/components/date-range/index.js]: Fixed both instance of not capitalising `And`

File: client/components/date-range/index.js [hunk 6]
```diff
@@ -153,6 +149,16 @@ export class DateRange extends Component {
 		}
 	};
 
+	closePopoverandRevert = () => {
```
[the remark this answers is not in the document]
Comment (getdave) [reply client/components/date-range/index.js]: I was thinking of them as an internal API method. They aren't necessarily bound to any particular scenario. You never know what scenarios in the future might require us to close and revert. Therefore I don't want to tie to any particular piece of UI or scenario. Rather make them clearly named and generic a la `closePopover`, `openPopover` and `togglePopover`.

````
DateRange. closePopoverandRevert()

DateRange. closePopoverandCommit()
```

It's clear what each does and for me it made the code much clearer.

**Pull request merged**